### PR TITLE
windows_certificate: properly add/remove pfx and private keys, changed output for all fetch commands to require an output path instead of dumping to screen, added the ability to create a certificate from a URL source

### PR DESCRIPTION
**Pull request by johnmccrae**
Signed-off-by: John McCrae <jmccrae@chef.io>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/resource/windows_certificate.rb [hunk 1]
```diff
@@ -92,16 +90,22 @@ class WindowsCertificate < Chef::Resource
         default: false,
         introduced: "16.8"
 
-      action :create, description: "Creates or updates a certificate" do
-        # Extension of the certificate
-        ext = ::File.extname(new_resource.source)
+      property :output_path, String,
+        description: "A path on the node where a certificate object (pfx, pem, cer, key, etc) can be exported to.",
+        introduced: "16.10"
```
Comment (tas50) [inline lib/chef/resource/windows_certificate.rb:95]: This will be 17.0 now

File: lib/chef/resource/windows_certificate.rb [hunk 2]
```diff
@@ -79,9 +80,6 @@ class WindowsCertificate < Chef::Resource
         description: "Use the `CurrentUser` store instead of the default `LocalMachine` store. Note: Prior to #{ChefUtils::Dist::Infra::CLIENT}. 16.10 this property was ignored.",
         default: false
 
-      property :cert_path, String,
```
Comment (tas50) [inline lib/chef/resource/windows_certificate.rb:82]: If this is being replaced with output path we need to do it in a backward-compatible way. https://docs.chef.io/dsl_custom_resource/#deprecate-and-alias

File: lib/chef/resource/windows_certificate.rb [hunk 3]
```diff
@@ -92,16 +90,22 @@ class WindowsCertificate < Chef::Resource
...
+        ext = get_file_extension(new_resource.source)
 
         # PFX certificates contains private keys and we import them with some other approach
         import_certificates(fetch_cert_object(ext), (ext == ".pfx"))
       end
 
       # acl_add is a modify-if-exists operation : not idempotent
-      action :acl_add, description: "Adds read-only entries to a certificate's private key ACL" do
+      action :acl_add do
+        description "Adds read-only entries to a certificate's private key ACL."
```
Comment (tas50) [inline lib/chef/resource/windows_certificate.rb:108]: This needs to be done the previous way. This way where description is its own method doesn't work

File: spec/functional/resource/windows_certificate2_spec.rb [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+      it "exports a PFX file with a valid thumbprint", :focus do
+        resource.source = tests_thumbprint
+        resource.pfx_password = password
+        resource.output_path = pfx_output_path
+        resource.run_action(:fetch)
+        expect(File.exist?(pfx_output_path)).to be_truthy
+      end
+
+      # it "exports a key file with a valid thumbprint" do
```
Comment (tas50) [inline spec/functional/resource/windows_certificate2_spec.rb:152]: We shouldn't have anything commented out

Comment (johnmccrae) [reply spec/functional/resource/windows_certificate2_spec.rb:152]: that file shouldn't even be there. UGH... freaking PR's.... that was a test file. Deleting it

File: chef-universal-mingw32.gemspec [hunk 5]
```diff
@@ -14,7 +14,7 @@ gemspec.add_dependency "win32-service", ">= 2.1.5", "< 3.0"
 gemspec.add_dependency "wmi-lite", "~> 1.0"
 gemspec.add_dependency "win32-taskscheduler", "~> 2.0"
 gemspec.add_dependency "iso8601", ">= 0.12.1", "< 0.14" # validate 0.14 when it comes out
-gemspec.add_dependency "win32-certstore", "~> 0.5.0" # 0.5+ required for specifying user vs. system store
+gemspec.add_dependency "win32-certstore", "~> 0.6.1" # 0.5+ required for specifying user vs. system store
```
Comment (tas50) [inline chef-universal-mingw32.gemspec:17]:
```suggestion
gemspec.add_dependency "win32-certstore", "~> 0.6.1" # 0.6+ required for user and system level cert handling
```

File: lib/chef/resource/windows_certificate.rb [hunk 6]
```diff
@@ -79,8 +80,7 @@ class WindowsCertificate < Chef::Resource
         description: "Use the `CurrentUser` store instead of the default `LocalMachine` store. Note: Prior to #{ChefUtils::Dist::Infra::CLIENT}. 16.10 this property was ignored.",
         default: false
 
-      property :cert_path, String,
-        description: "The path to the certificate."
+      deprecated_property_alias :cert_path, :output_path, "The cert_path property was renamed output_path in the 17.0 release of this cookbook. Please update your cookbooks to use the new property name."
```
Comment (tas50) [inline lib/chef/resource/windows_certificate.rb:83]:
```suggestion
      deprecated_property_alias :cert_path, :output_path, "The cert_path property was renamed output_path in the 17.0 release of #{ChefUtils::Dist::Infra::CLIENT}. Please update your cookbooks to use the new property name."
```

File: lib/chef/resource/windows_certificate.rb [hunk 7]
```diff
@@ -62,11 +63,11 @@ class WindowsCertificate < Chef::Resource
       DOC
 
       property :source, String,
-        description: "The source file (for create and acl_add), thumbprint (for delete and acl_add) or subject (for delete) if it differs from the resource block's name.",
+        description: "The source file (for create and acl_add), thumbprint (for delete, export and acl_add) or subject (for delete or export) if it differs from the resource block's name.",
```
Comment (IanMadd) [inline lib/chef/resource/windows_certificate.rb:66]:
```suggestion
        description: "The source file (for `create` and `acl_add`), thumbprint (for `delete`, `export`, and `acl_add`), or subject (for `delete` or `export`) if it differs from the resource block's name.",
```

File: lib/chef/resource/windows_certificate.rb [hunk 8]
```diff
@@ -62,11 +63,11 @@ class WindowsCertificate < Chef::Resource
...
         name_property: true
 
       property :pfx_password, String,
-        description: "The password to access the source if it is a pfx file."
+        description: "The password to access the object with if it is a pfx file."
```
Comment (IanMadd) [inline lib/chef/resource/windows_certificate.rb:70]:
```suggestion
        description: "The password to access the object with if it is a PFX file."
```

File: lib/chef/resource/windows_certificate.rb [hunk 9]
```diff
@@ -92,16 +92,19 @@ class WindowsCertificate < Chef::Resource
         default: false,
         introduced: "16.8"
 
-      action :create, description: "Creates or updates a certificate" do
-        # Extension of the certificate
-        ext = ::File.extname(new_resource.source)
+      property :output_path, String,
+        description: "A path on the node where a certificate object (pfx, pem, cer, key, etc) can be exported to.",
```
Comment (IanMadd) [inline lib/chef/resource/windows_certificate.rb:96]:
```suggestion
        description: "A path on the node where a certificate object (PFX, PEM, CER, KEY, etc) can be exported to.",
```

**Comment by johnmccrae** [comment]
This PR was totally busted. It was replaced by #11405

**Pull request closed without merging**